lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -965,41 +965,77 @@ class ADContext {
-      // capacity than the desired indices. We expect this logic to go away when
-      // we support `@differentiable` partial apply.
-      if (rdaIndexSet->isSupersetOf(
-              indexSet->extendingCapacity(getASTContext(),
-                                          rdaIndexSet->getCapacity())) &&
+      // NOTE(TF-642): `da` may come from a un-partial-applied function and
+      // have larger capacity than the desired indices. We expect this logic to
+      // go away when `partial_apply` supports `@differentiable` callees.
+      if (daIndexSet->isSupersetOf(indexSet->extendingCapacity(
+              getASTContext(), daIndexSet->getCapacity())) &&
           // fewer parameters than before
-          (superset->isEmpty() ||
-           rdaIndexSet->getNumIndices() < superset->getNumIndices())) {
-        superset = rda->getIndices().parameters;
-        minimalAttr = rda;
+          (minimalIndexSet->isEmpty() ||
+           daIndexSet->getNumIndices() < minimalIndexSet->getNumIndices())) {
+        minimalAttr = da;
+        minimalIndexSet = daIndexSet;
       }
     }
     return minimalAttr;
   }
 
+  /// Finds the `@differentiable` attribute (and its parameter indices) on the
+  /// specified original function whose parameter indices are a minimal
+  /// superset of the specified parameter indices. Returns nullptr if no such
+  /// attribute exists.
+  std::pair<const DifferentiableAttr *, AutoDiffIndexSubset *>
+  lookUpMinimalASTDifferentiableAttrAndIndexSubset(
+      SILDeclRef originalDeclRef, CanSILFunctionType originalFnType,
+      const SILAutoDiffIndices &indices) {
+    auto *original = originalDeclRef.getDecl();
+    const DifferentiableAttr *minimalAttr = nullptr;
+    auto *minimalIndexSet = AutoDiffIndexSubset::getDefault(
+        getASTContext(), originalFnType->getNumParameters(), false);
+    auto *indexSet = indices.parameters;
+    for (auto *da : original->getAttrs().getAttributes<DifferentiableAttr>()) {
+      auto *daParamIndices = da->getParameterIndices();
+      auto *daIndexSet = daParamIndices->getLowered(
+          getASTContext(),
+          original->getInterfaceType()->castTo<AnyFunctionType>());
+      // If all indices in `indexSet` are in `daIndexSet`, and it has fewer
+      // indices than our current candidate and a primitive VJP, then `da` is
+      // our new candidate.
+      //
+      // NOTE(TF-642): `da` may come from a un-partial-applied function and
+      // have larger capacity than the desired indices. We expect this logic to
+      // go away when `partial_apply` supports `@differentiable` callees.
+      if (daIndexSet->isSupersetOf(indexSet->extendingCapacity(getASTContext(),
+              daIndexSet->getCapacity())) &&
+          // fewer parameters than before
+          (minimalIndexSet->isEmpty() ||
+           daIndexSet->getNumIndices() < minimalIndexSet->getNumIndices())) {
+            minimalAttr = da;
```

Comment:
Fix indentation.

Reply:
Thanks, done in https://github.com/apple/swift/pull/26313/commits/1e535fa27077fb836653365b953b7c564e5dbf58.